### ISSUE TITLE
Improvements/User experience : Over upgrade/install versions > 0.9.0 prerequisites

### ISSUE THREAD

**Issue by camilamacedo86** [operator-framework/operator-sdk#1732]
## Feature Request
- Add info the [prerequisites](https://github.com/operator-framework/operator-sdk/blob/21668facf7a72fa7cb49bd087eaab0ee103f51b0/README.md#prerequisites) that for the versions > 0.9.0 is needed  `GO111MODULE=on environment variable set.`
- Add info the [prerequisites](https://github.com/operator-framework/operator-sdk/blob/21668facf7a72fa7cb49bd087eaab0ee103f51b0/README.md#prerequisites) for versions > 0.9.0 over [mercurial v3.9+](https://www.mercurial-scm.org/wiki/Download) 
- Improve the error message [here](
https://github.com/operator-framework/operator-sdk/blob/907c4a7b9164f6d33530a91797f9272c431df8fc/cmd/operator-sdk/main.go#L147). In order to provide a better experience, we can be more clear over it and give the step required for the user to solve it. 

NOTE: Users which will upgrade the version to 0.9.0 will face it. 

Following an example. 

```shell
 $ brew upgrade operator-sdk
==> Upgrading 1 outdated package:
operator-sdk 0.8.1 -> 0.9.0
==> Upgrading operator-sdk 
==> Installing dependencies for operator-sdk: go
==> Installing operator-sdk dependency: go
==> Downloading https://homebrew.bintray.com/bottles/go-1.12.7.mojave.bottle.tar.gz
Already downloaded: /Users/camilamacedo/Library/Caches/Homebrew/downloads/004ca83bdc17ddc6acaf8f8deef804dacb23782c1ba0b20a72a98bb1eb9e1efa--go-1.12.7.mojave.bottle.tar.gz
==> Pouring go-1.12.7.mojave.bottle.tar.gz
🍺  /usr/local/Cellar/go/1.12.7: 9,816 files, 452.7MB
==> Installing operator-sdk
==> Downloading https://homebrew.bintray.com/bottles/operator-sdk-0.9.0.mojave.bottle.tar.gz
==> Downloading from https://akamai.bintray.com/33/33202cb5bc7c7b3cd1346dc3a1f1ed982eb8676e5a073d21096af80d5a81efcd?__gda__=exp=1563971719~hmac=c72dfda7d3ea6574481b823abdfb45181ceffb236131700d900b41e42e02d60e&response-content-disposition=attachmen
######################################################################## 100.0%
==> Pouring operator-sdk-0.9.0.mojave.bottle.tar.gz
==> Caveats
Bash completion has been installed to:
  /usr/local/etc/bash_completion.d

zsh completions have been installed to:
  /usr/local/share/zsh/site-functions
==> Summary
🍺  /usr/local/Cellar/operator-sdk/0.9.0: 8 files, 63.6MB
Removing: /usr/local/Cellar/operator-sdk/0.8.1... (8 files, 68.1MB)
Removing: /Users/camilamacedo/Library/Caches/Homebrew/operator-sdk--0.8.1.mojave.bottle.tar.gz... (26.4MB)
==> Caveats
==> operator-sdk
Bash completion has been installed to:
  /usr/local/etc/bash_completion.d

zsh completions have been installed to:
  /usr/local/share/zsh/site-functions
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/test $ operator-sdk new app-test
INFO[0000] Creating new Go operator 'app-test'.         
Error: dependency manager "modules" requires working directory to be in $GOPATH/src and GO111MODULE=on, or outside of $GOPATH/src and GO111MODULE="on", "auto", or unset
Usage:
  operator-sdk new <project-name> [flags]

Flags:
      --api-version string          Kubernetes apiVersion and has a format of $GROUP_NAME/$VERSION (e.g app.example.com/v1alpha1) - used with "ansible" or "helm" types
      --dep-manager string          Dependency manager the new project will use (choices: "dep", "modules") (default "modules")
      --generate-playbook           Generate a playbook skeleton. (Only used for --type ansible)
      --git-init                    Initialize the project directory as a git repository (default false)
      --header-file string          Path to file containing headers for generated Go files. Copied to hack/boilerplate.go.txt
      --helm-chart string           Initialize helm operator with existing helm chart (<URL>, <repo>/<name>, or local path)
      --helm-chart-repo string      Chart repository URL for the requested helm chart
      --helm-chart-version string   Specific version of the helm chart (default is latest version)
  -h, --help                        help for new
      --kind string                 Kubernetes CustomResourceDefintion kind. (e.g AppService) - used with "ansible" or "helm" types
      --repo string                 Project repository path for Go operators. Used as the project's Go import path. This must be set if outside of $GOPATH/src with Go modules, and cannot be set if --dep-manager=dep
      --skip-validation             Do not validate the resulting project's structure and dependencies. (Only used for --type go)
      --type string                 Type of operator to initialize (choices: "go", "ansible" or "helm") (default "go")
      --vendor                      Use a vendor directory for dependencies. This flag only applies when --dep-manager=modules (the default)

Global Flags:
      --verbose   Enable verbose logging

camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/test $ export GO111MODULE=on
```

**Comment by hasbro17** [comment]
> Add info the prerequisites that for the versions > 0.9.0 is needed GO111MODULE=on environment variable set.

You don't always need to set `GO111MODULE=on` as prerequisite. You can be using the operator-sdk cmd outside of `$GOPATH/src` where it's not necessary to set `GO111MODULE=on`.
That's why we have a specific section explaining the two scenarios in the user-guide.
https://github.com/operator-framework/operator-sdk/blob/master/doc/user-guide.md#go-modules   


> Add info the prerequisites for versions > 0.9.0 over mercurial v3.9+

We're already listing `mercurial version 3.9+` in the [prerequisites](https://github.com/operator-framework/operator-sdk#prerequisites). Can you clarify what more info you meant with regards to this?

> Improve the error message [here](https://github.com/operator-framework/operator-sdk/blob/907c4a7b9164f6d33530a91797f9272c431df8fc/cmd/operator-sdk/main.go#L147). In order to provide a better experience, we can be more clear over it and give the step required for the user to solve it.

The current error message is supposed to tell the users that Go modules is not enabled and there's two ways to use it.
```console
Error: dependency manager "modules" requires working directory to be in $GOPATH/src and GO111MODULE=on, or outside of $GOPATH/src and GO111MODULE="on", "auto", or unset
```

Essentially mirroring the Go Modules wiki [instructions for enabling Go Modules](https://github.com/golang/go/wiki/Modules#how-to-install-and-activate-module-support):
```console
- Invoke the go command in a directory outside of the $GOPATH/src tree, with a valid go.mod file in the current directory or any parent of it and the environment variable GO111MODULE unset (or explicitly set to auto).
- Invoke the go command with GO111MODULE=on environment variable set.
```

Can you elaborate on what's ambiguous in the current error message?
Maybe we can rephrase it to be have more explicit steps e.g:
```console
Error: dependency manager "modules" requires Go modules to be enabled. Enable it in one of the following ways:
- Invoke the `operator-sdk new` command in a directory outside of the $GOPATH/src tree, with environment variable GO111MODULE unset (or explicitly set to auto)
- Invoke the `operator-sdk new` command inside $GOPATH/src with GO111MODULE=on environment variable set.
```

Feel free to submit a PR for that.

**Comment by camilamacedo86** [comment]
HI @hasbro17, 

Thank you for your reply. Following the comments inline regards the above points. 

## Regards `GO111MODULE=on` info. 

- Note that in we guide the new users to create use the commands inside of `$GOPATH/src`  then IMHO shows that the beginner's ones will face this issue. 
- Could we add some reference in the install guide to address the user to check if it: https://github.com/operator-framework/operator-sdk/blob/master/doc/user-guide.md#go-modules? 

The point here is trying to makes this info easier for the new users in order to avoid them face this scenario with the latest version.  

## Regards `mercurial version 3.9+ in the prerequisites` 
It can be ignored since it is already done. Sorry, probably I was not checking it in the master branch.  

## Regards the error msg `Error: dependency manager "modules"   ...`

IMHO: we could improve it as you suggested as well and make more clear the steps required to fix it. Your suggestion would make it great already. 

> Error: dependency manager "modules" requires Go modules to be enabled. Enable it in one of the following ways:
> - Invoke the `operator-sdk new` command in a directory outside of the $GOPATH/src tree, with environment variable GO111MODULE unset (or explicitly set to auto)
> - Invoke the `operator-sdk new` command inside $GOPATH/src with GO111MODULE=on environment variable set.

**Comment by camilamacedo86** [comment]
Closing since we could check that the info is added in the current version of the docs and the #1776 is already addressing the improve in the msg which was its goal.
